geopandas/io/tests/test_file.py

```diff
@@ -139,15 +139,31 @@ def test_to_file_bool(tmpdir, driver, ext):
     assert_correct_driver(tempfilename, ext)
 
 
-def test_to_file_datetime(tmpdir):
+dt_resolution = ["ms", "s", "min", "h", "D"]
+
+
+@pytest.mark.parametrize("dt_res", dt_resolution)
+@pytest.mark.parametrize("driver,ext", driver_ext_pairs)
+def test_to_file_datetime(tmpdir, driver, ext, dt_res):
     """Test writing a data file with the datetime column type"""
-    tempfilename = os.path.join(str(tmpdir), "test_datetime.gpkg")
+    if ext in (".shp", ""):
+        pytest.skip(f"Driver corresponding to ext {ext} doesn't support dt fields")
+    tempfilename = os.path.join(str(tmpdir), f"test_datetime{ext}")
     point = Point(0, 0)
     now = datetime.datetime.now()
     df = GeoDataFrame({"a": [1, 2], "b": [now, now]}, geometry=[point, point], crs=4326)
-    df.to_file(tempfilename, driver="GPKG")
+    if compat.FIONA_GE_1814:
+        fiona_precision_limit = "ms"
+    else:
+        fiona_precision_limit = "s"
+    df["b"] = df["b"].dt.round(freq=fiona_precision_limit)
```

Comment:
I've cleaned this up and the dodgy tolerance checking I was doing. Looking at this some more, the "random" error was just due to rounding of microseconds (which was random because `datetime.datetime.now()` is in the test). Fiona was rounding them and the expected case I was comparing to was just truncating. I think this is clearer, but now think that the parametersising on time resolution is a bit pointless. If they should stay, then it probably make sense to do something better for the fiona <1.8.14 case, as ms precision passes because it's rounded before that point.

There is also perhaps a separate question about whether there should be any warning about dropping microseconds/ nanoseconds, but this would have been happening prior to this PR anyway.

Reply:
What was the initial goal of testing with multiple resolutions? Fiona just seems to have "datetime" handling, and not actually handling if different resolutions? (or do some formats infer that?)

Reply:
To be honest, I'm not sure what I was thinking at the time, probably thinking that date was potentially distinct to datetime and got carried away. But as you say - and as I realised when coming back to this, if it's all handled as datetimes in fiona then this doesn't make any difference, so I'll drop that out.